crates/red_knot_test/README.md

```diff
@@ -0,0 +1,430 @@
+# Writing type-checking / type-inference tests
+
+Any Markdown file can be a test suite.
+
+In order for it to be run as one, `red_knot_test::run` must be called with its path; see
+`crates/red_knot_python_semantic/tests/mdtest.rs` for an example that treats all Markdown files
+under a certain directory as test suites.
+
+A Markdown test suite can contain any number of tests. A test consists of one or more embedded
+"files", each defined by a triple-backticks fenced code block. The code block must have a tag string
+specifying its language; currently only `py` (Python files) and `pyi` (type stub files) are
+supported.
+
+The simplest possible test suite consists of just a single test, with a single file.
+
+````markdown
+```py
+reveal_type(1)  # revealed: Literal[1]
+```
+````
+
+When running this test, the mdtest framework will write a file with these contents to the default
+file path (`/src/test.py`) in its in-memory file system, run a type check on that file, and then
+match the resulting diagnostics with the assertions in the test. Assertions are in the form of
+Python comments. If all diagnostics and all assertions are matched, the test passes; otherwise, it
+fails.
+
+## Assertions
+
+Two kinds of assertions are supported, `# revealed:` (shown above) and `# error:`.
+
+### Assertion kinds
+
+#### revealed
+
+A `# revealed:` assertion should always be paired with a call to the `reveal_type` utility, which
+reveals (via a diagnostic) the inferred type of its argument (which can be any expression). The text
+after `# revealed:` must match exactly with the displayed form of the revealed type of that
+expression.
+
+The `reveal_type` function can be imported from the `typing` standard library module (or, for older
+Python versions, from the `typing_extensions` pseudo-standard-library module -- it is a third-party
+module, but typeshed, and thus type checkers also, treat it as part of the standard library):
```

Comment:
Using a footnote for the digression is a good idea, I'll do that.

> I'd wager that 99% of the time you'll want to import it from `typing_extensions` rather than `typign`

I don't think this is true; I think rather that 99% of tests will not import it at all, and the only ones that do will be a few tests that are explicitly testing the handling of imported vs pseudo-builtin `reveal_type`, and of those few tests probably roughly half (i.e. just one or two tests each) will import it each way.

I think the more important consideration here is that the preferred and modern thing to do, if on a Python version where it's possible, is to import from `typing`, so that's what should be featured, and the legacy approach should be a parenthetical.